### DISCUSSION TITLE
SSR with a backend other than Node.js

### DISCUSSION THREAD

**Discussion by babgyy** [algolia/instantsearch#5283] [Ideas]
I use React-Instantsearch with a Ruby on Rails backend. I would like to have the initial view rendered server-side for SEO reasons. But I can't find a way to make it work.

I can easily make an Algolia query server-side in ruby, get a JSON payload from the `algoliasearch-rails` gem with the `raw_result` method, however this does not work when given as a `resultsState` prop to InstantSearch. 

Somebody had this [same question](https://github.com/algolia/react-instantsearch/issues/649) 2 years ago. But I feel a lot of work has been done by Algolia since then to make this possible. Is it ? 

**How would you suggest doing SSR without** 

- using Node as a backend
- switching entirely to a custom server endpoint ? 

**Is there a way to use `findResultsState` in my context ?** 

**Is there any clear schema documentation for the `resultsState` prop ?**

I have read InstantSearch's documentation of [resultsState props](https://www.algolia.com/doc/api-reference/widgets/instantsearch/react/?language=react#widget-param-resultsstate) that recommends to use `findResultsState`. Does this method have a Ruby / Rails equivalent ? 

I have read the [Server side rendering guide ](https://www.algolia.com/doc/guides/building-search-ui/going-further/server-side-rendering/react/) but it assumes I have a JS backend.

Any directions would be appreciated

**Comment by Haroenv** [comment]
This seems like a quite hard question to answer directly. Could you make a setup we can try out things in? 

What kind of setup do you have to eventually render the react app?

**Comment by babgyy** [comment]
I have [`webpacker`](https://github.com/rails/webpacker) (webpack 3 preconfigured for rails common usecases) 

React components are rendered with [`react-rails`](https://github.com/reactjs/react-rails). This library consists mainly of ruby helper methods to trigger the mounting of React components on page loads, and to enable Server Side Rendering with [`ExecJS`](https://github.com/rails/execjs).

I will create a setup right now.

**Comment by Haroenv** [comment]
Since ExecJS doesn't support modules, I'd be surprised if it's possible this way, but interested to see what you have so far

**Comment by babgyy** [comment]
Here is a demo setup : https://github.com/babgyy/ssr-instantsearch-rails

I documented a bit how to setup and how to make it fail. As soon as I added a facet search, the feeding of React Instantsearch with JSON results from ruby began to fail. Activating SSR just produces the error sooner but fails the same way.

> Since ExecJS doesn't support modules, I'd be surprised if it's possible this way, but interested to see what you have so far

In my understanding, Webpack & Babel handle modules and produce a JS edible by ExecJS

**Comment by babgyy** [comment]
Any opinion on this ?

**Comment by Haroenv** [comment]
Can you show _where_ in that repository the SSR is actually happening? 

Note that for `resultsState` to work, you'll need to make `findResultsState` (from SSR) work, or replicate it in the ruby part, which won't be easy at all, since a refinement list for example has multiple requests that are done. For that the algolia JS helper is used.

**Comment by babgyy** [comment]
> Can you show where in that repository the SSR is actually happening?

Sure, as stated in the [Argument section](https://github.com/babgyy/ssr-instantsearch-rails#argument), SSR is handled by the `react-rails` gem. As a user of the gem, I "just" pass `{prerendering: true}` as an argument to the `react_component` ruby helper in the view `app/views/welcome/index.html.erb`

For more details on how this is all passed to ExecJS, we would have to drill into the implementation of the gem. Not uninteresting but irrelevant to this question in my opinion

> Note that for `resultsState` to work, you'll need to make `findResultsState` (from SSR) work, or replicate it in the ruby part, which won't be easy at all, since a refinement list for example has multiple requests that are done. For that the algolia JS helper is used.

Precisely my point :) I understand that `resultsState` requires `findResultsState`. Which raises the following questions : 

1. is it possible to feed the InstantSearch with a JSON payload of results ?
2. if not, how would you recommend doing SSR of `InstantSearch` with any other backend than a Node/JS backend ?

**Comment by Haroenv** [comment]
I recommend using node for server side rendering in general, you'll have a lot less headaches, and will always fit in the "simple path" of server side rendering. You can set up the server for server side rendering separately from your other backend easily.
